hubspot-style/src/main/java/com/hubspot/immutables/utils/WireSafeEnum.java

```diff
@@ -295,32 +302,46 @@ public JsonDeserializer<?> createContextual(DeserializationContext ctxt, BeanPro
         } else if (!typeParameters[0].isEnumType()) {
           throw ctxt.mappingException("Can not handle non-enum type: " + typeParameters[0].getRawClass());
         } else {
-          return deserializerFor(typeParameters[0].getRawClass());
+          return deserializerFor(typeParameters[0]);
         }
       }
     }
 
-    @SuppressWarnings("unchecked")
-    private static <T extends Enum<T>> JsonDeserializer<?> deserializerFor(Class<?> rawType) {
-      return DESERIALIZER_CACHE.computeIfAbsent(rawType, ignored -> {
-        Class<T> enumType = (Class<T>) rawType;
-        return newDeserializer(enumType);
-      });
+    private static JsonDeserializer<?> deserializerFor(JavaType javaType) {
+      return DESERIALIZER_CACHE.computeIfAbsent(javaType, Deserializer::newDeserializer);
     }
 
-    private static <T extends Enum<T>> JsonDeserializer<WireSafeEnum<?>> newDeserializer(Class<T> enumType) {
+    private static <T extends Enum<T>> JsonDeserializer<WireSafeEnum<?>> newDeserializer(JavaType enumType) {
       return new JsonDeserializer<WireSafeEnum<?>>() {
 
         @Override
         public WireSafeEnum<T> deserialize(JsonParser p, DeserializationContext ctxt) throws IOException {
           if (p.getCurrentToken() == JsonToken.VALUE_NULL) {
             return null;
           } else if (p.getCurrentToken() == JsonToken.VALUE_STRING) {
-            return WireSafeEnum.fromJson(enumType, p.getText());
+            @SuppressWarnings("unchecked")
+            Class<T> rawType = (Class<T>) enumType.getRawClass();
+            return WireSafeEnum.fromJson(rawType, p.getText(), (klass, value) -> create(klass, value, p, ctxt));
           } else {
             throw ctxt.wrongTokenException(p, JsonToken.VALUE_STRING, null);
           }
         }
+
+        private WireSafeEnum<T> create(Class<T> rawClass, String jsonValue, JsonParser parser, DeserializationContext ctxt) {
+          return deserializeValue(parser, ctxt)
+              .map(v -> new WireSafeEnum<>(rawClass, jsonValue, v))
+              .orElseGet(() -> new WireSafeEnum<>(rawClass, jsonValue));
+        }
+
+        @SuppressWarnings("unchecked")
+        private Optional<T> deserializeValue(JsonParser p, DeserializationContext ctxt) {
+          try {
+            JsonDeserializer<?> deserializer = ctxt.findNonContextualValueDeserializer(enumType);
```

Comment:
is the `findNonContextualValueDeserializer` the mechanism by which jackson's @JsonCreator machinery gets invoked here?

Reply:
Yea, by delegating here we call back into the object mapper asking for it to figure out whatever is the appropriate deserializer to use, which allows us to leverage the normal jackson annotations / binding machinery